### PR TITLE
storage: create new tracing span for each replicatedCmd during application

### PR DESCRIPTION
**Pull request by nvb**
This change adjusts our handling of contexts and replicatedCmds. We no longer
assign a local proposal's context to its replicatedCmds directly. Instead,
we always create a new tracing span that follows from this context. This
eliminates a whole class of bugs that we have fought to fix in changes
like #39203. In fact, we are still seeing issues after the recent refactor
when stressing #39390.

The change also paves the way for tracing the application of command application
on follower replicas.

### CODE REVIEW

**Review by andreimatei — APPROVED** [review]
LGTM

> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39425#-:-Lli6zym2If2q266bUh1:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/storage/replica_application_decoder.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39425#-Lli6tYe9zjHEYjRvH-g:-Lli6tYfC4O2hOJH8jXu:b-18bwix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8b047b32ea6e42108453faa0f598a5f5f05213/pkg/storage/replica_application_decoder.go#L104)):*
> ```Go
> 			// proposal from the map. We expect this entry to be rejected by
> 			// checkForcedErr.
> 			cmd.raftCmd.MaxLeaseIndex == cmd.proposal.command.MaxLeaseIndex
> ```

so this is... a fix unrelated to the tracing changes?
___
*[pkg/storage/replica_application_state_machine.go, line 854 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39425#-Lli5vzL4Fkz5afu-M1C:-Lli5vzM5x7GSkcFCWa2:b-54dabk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9990dc7d7ec7459b4da8ce27bb67e8f8f1fc985f/pkg/storage/replica_application_state_machine.go#L854)):*
> ```Go
> 				log.Fatalf(ctx, "command already applied: %+v; unexpected successful result", cmd)
> 			}
> 			cmd.proposal = nil
> ```

how come this went away?

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39425)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/storage/replica_application_decoder.go, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39425#-Lli6tYe9zjHEYjRvH-g:-Lli7ep32tJbhO7SQYYs:b6ebnfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8b047b32ea6e42108453faa0f598a5f5f05213/pkg/storage/replica_application_decoder.go#L104)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

so this is... a fix unrelated to the tracing changes?
</blockquote></details>

Well, it was a workaround that we got to remove now that we don't have to worry about use-after-free bugs. Forking the context span fixes all of these kinds of issues.
___
*[pkg/storage/replica_application_state_machine.go, line 854 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39425#-Lli5vzL4Fkz5afu-M1C:-Lli85-8EaFgDTNATLRm:b-883ab5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9990dc7d7ec7459b4da8ce27bb67e8f8f1fc985f/pkg/storage/replica_application_state_machine.go#L854)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how come this went away?
</blockquote></details>

Because we weren't allowed to call `ProposalData.finishApplication` multiple times. Doing this wasn't really a great way to approach the problem.

File: pkg/storage/replica_application_cmd.go [hunk 1]
```diff
@@ -112,10 +118,10 @@ func (c *replicatedCmd) Rejected() bool {
 
 // FinishAndAckOutcome implements the apply.AppliedCommand interface.
 func (c *replicatedCmd) FinishAndAckOutcome() error {
```
Comment (tbg) [inline pkg/storage/replica_application_cmd.go]: It's documented that this must only be called once, right?

**Comment by tbg** [comment]
> In fact, we are still seeing issues after the recent refactor
when stressing #39390.

Curious whether we're just throwing the towel because it's not worth to fix it, or whether you found something that means that we would never be able to fix it.

**Comment by nvanbenschoten** [comment]
> Curious whether we're just throwing the towel because it's not worth to fix it, or whether you found something that means that we would never be able to fix it.

I had a different fix for this but it was going to make things even harder on @andreimatei in https://github.com/cockroachdb/cockroach/pull/39431. Talking to him set me straight and made me realize that this is free in the common case where there is no tracing span already in the proposal's context and it avoids the entire class of bugs that we had been seeing.

I consider this the correct fix given the lifetimes of these `replicatedCmd`s and how they interact with the lifetime of the proposer's context. That's especially true once we start acking clients before entry application more frequently.

File: pkg/storage/replica_application_cmd.go (hunk 1 above)
Comment (nvanbenschoten) [reply pkg/storage/replica_application_cmd.go]: It is in https://github.com/cockroachdb/cockroach/pull/39468.

**Pull request merged**